### PR TITLE
Add a persistent layer of setting options and introduce a Tutorial mode

### PR DESCRIPTION
**Pull request by kellertuer**
This adresses a point raised in #346 and is for now realised in the `quasi_Newton` solver.

* [x] check which other solvers benefit from the “Tutorial” and the new warning
* [x] check test coverage
* [x] document the new mode
* [x] illustrate / mention the new mode in the intro tutorial.

### CODE REVIEW

**Comment by kellertuer** [comment]
So the tutorial already has a section about this, I just have to figure out how to convince Quarto to display warnings :)

https://manoptjl.org/previews/PR354/tutorials/Optimize/#Using-the-“Tutorial”-mode

File: src/plans/debug.jl [hunk 1]
```diff
@@ -1028,6 +1082,7 @@ Note that the Shortcut symbols should all start with a capital letter.
 
 * `:Cost` creates a [`DebugCost`](@ref)
 * `:Change` creates a [`DebugChange`](@ref)
+* `:Gradient` creattes a [`DebugGradient`](@ref)
```
Comment (mateuszbaran) [inline src/plans/debug.jl:1085]:
```suggestion
* `:Gradient` creates a [`DebugGradient`](@ref)
```

File: src/plans/plan.jl [hunk 2]
```diff
@@ -25,6 +25,8 @@ end
 """
     get_manopt_parameter(f, element::Symbol, args...)
 
+Access arbitrary parameters from `f` adressedby a symbol `element`.
```
Comment (mateuszbaran) [inline src/plans/plan.jl:28]:
```suggestion
Access arbitrary parameters from `f` addressed by a symbol `element`.
```

File: src/plans/plan.jl [hunk 3]
```diff
@@ -25,6 +25,8 @@ end
...
+
 For any `f` and a `Symbol` `e` we dispatch on its value so by default, to
 get some element from `f` potentially further qulalified by `args...`.
```
Comment (mateuszbaran) [inline src/plans/plan.jl:31]:
```suggestion
get some element from `f` potentially further qualified by `args...`.
```

File: src/plans/plan.jl [hunk 4]
```diff
@@ -35,6 +37,57 @@ function get_manopt_parameter(f, e::Symbol, args...)
...
+
+The parameters are queried from the global settings using [`Preferences.jl`](https://github.com/JuliaPackaging/Preferences.jl),
+so they are persistent within your activated Environment.
+
+# Currenlty used settings
+
+`:Mode`
+the mode can be set to `"Tutorial"` to get several hints especially in scenarios, where
+the optimisation on manifolds is different from the usual “experience” in
+(classicall, Euclidean) optimization.
```
Comment (mateuszbaran) [inline src/plans/plan.jl:56]:
```suggestion
(classical, Euclidean) optimization.
```

File: src/plans/plan.jl [hunk 5]
```diff
@@ -35,6 +37,57 @@ function get_manopt_parameter(f, e::Symbol, args...)
...
+
+Access global [`Manopt`](@ref) parametersadressed by a symbol `element`.
+We first dispatch on the value of `element`.
+
+if the value is not set, `default` is returned
+
+The parameters are queried from the global settings using [`Preferences.jl`](https://github.com/JuliaPackaging/Preferences.jl),
+so they are persistent within your activated Environment.
+
+# Currenlty used settings
```
Comment (mateuszbaran) [inline src/plans/plan.jl:51]:
```suggestion
# Currently used settings
```

File: src/plans/plan.jl [hunk 6]
```diff
@@ -35,6 +37,57 @@ function get_manopt_parameter(f, e::Symbol, args...)
...
+    e::Symbol, args...; default=get_manopt_parameter(Val(e), Val(:default))
+)
+    return @load_preference("$(e)", default)
+end
+# Handle empty defaults
+get_manopt_parameter(::Symbol, ::Val{:default}) = nothing
+get_manopt_parameter(::Val{:Mode}, v::Val{:default}) = ""
+
+"""
+    set_manopt_parameter!(element::Symbol, value::Union{String,Bool})
```
Comment (mateuszbaran) [inline src/plans/plan.jl:69]:
```suggestion
    set_manopt_parameter!(element::Symbol, value::Union{String,Bool,<:Number})
```

File: src/plans/plan.jl [hunk 7]
```diff
@@ -35,6 +37,57 @@ function get_manopt_parameter(f, e::Symbol, args...)
...
+
+Set global [`Manopt`](@ref) parameters adressed by a symbol `element`.
+We first dispatch on the value of `element`.
+
+The parameters are stored to the global settings using [`Preferences.jl`](https://github.com/JuliaPackaging/Preferences.jl).
+
+Passing a `value` of `""` deletes an entry from the preferences.
```
Comment (mateuszbaran) [inline src/plans/plan.jl:76]:
```suggestion
Passing a `value` of `""` deletes the corresponding entry from the preferences.
```

File: src/plans/plan.jl [hunk 8]
```diff
@@ -35,6 +37,57 @@ function get_manopt_parameter(f, e::Symbol, args...)
...
+Whenever the `LocalPreferences.toml` is modified, this is also `@info`rmed about.
+"""
+function set_manopt_parameter!(e::Symbol, value::Union{String,Bool,<:Number})
+    if length(value) == 0
+        @delete_preferences!("$(e)")
+        v = get_manopt_parameter(e, Val(:default))
+        default = isnothing(v) ? "" : (length(v) == 0 ? "" : " ($(get_manopt_parameter))")
```
Comment (mateuszbaran) [inline src/plans/plan.jl:83]: Is this really supposed to just print `get_manopt_parameter` without any arguments?

File: src/plans/plan.jl [hunk 9]
```diff
@@ -35,6 +37,57 @@ function get_manopt_parameter(f, e::Symbol, args...)
...
+We first dispatch on the value of `element`.
+
+The parameters are stored to the global settings using [`Preferences.jl`](https://github.com/JuliaPackaging/Preferences.jl).
+
+Passing a `value` of `""` deletes an entry from the preferences.
+Whenever the `LocalPreferences.toml` is modified, this is also `@info`rmed about.
+"""
+function set_manopt_parameter!(e::Symbol, value::Union{String,Bool,<:Number})
+    if length(value) == 0
+        @delete_preferences!("$(e)")
```
Comment (mateuszbaran) [inline src/plans/plan.jl:81]:
```suggestion
        @delete_preferences!(string(e))
```
This is, I think, a more standard way of converting symbols to strings.

File: src/plans/stopping_criterion.jl [hunk 10]
```diff
@@ -685,7 +634,7 @@ A functor for an stopping criterion, where the algorithm if stopped when a varia
 # Fields
 * `value` – stores the variable which has to fall under a threshold for the algorithm to stop
 * `minValue` – stores the threshold where, if the value is smaller or equal to this threshold, the algorithm stops
-* `reason` – stores a reason of stopping if the stopping criterion has one be
+* `reason` – stores a reason of stopping if the stopping criteri on has one be
```
Comment (mateuszbaran) [inline src/plans/stopping_criterion.jl:637]: ?

File: src/solvers/gradient_descent.jl [hunk 11]
```diff
@@ -240,7 +240,15 @@ function gradient_descent!(
     ),
     stopping_criterion::StoppingCriterion=StopAfterIteration(200) |
                                           StopWhenGradientNormLess(1e-8),
-    debug=stepsize isa ConstantStepsize ? [DebugWarnIfCostIncreases()] : [],
+    debug=if (get_manopt_parameter(:Mode) == "Tutorial")
```
Comment (mateuszbaran) [inline src/solvers/gradient_descent.jl:243]: This is so much logic that I would suggest making a separate function for computing this default.

Comment (mateuszbaran) [reply src/solvers/gradient_descent.jl:243]: BTW, I'd also suggest making a helper `is_tutorial_mode` function.

File: tutorials/Optimize.qmd [hunk 12]
```diff
@@ -156,9 +155,62 @@ Then we reach approximately the same point as in the previous run, but in far le
...
+grad_f2(M, p) = sum(grad_distance.(Ref(M), data, Ref(p)));
+```
+
+And we consider the gradient at the starting point in norm
+
+```{julia}
+norm(M, data[1], grad_f2(M, data[1]))
+```
+
+On the sphere, when we follow a geodesic, we “return” to the start point after length $2π$. If we “lang” short before the starting point due to a gradient of length just shy of $2π$, line search would not make much sense probably.
```
Comment (mateuszbaran) [inline tutorials/Optimize.qmd:184]: What do you mean by "lang"?

**Review by mateuszbaran — COMMENTED** [review]
The suggested usage of Preferences.jl is that you restart your Julia session to take any preference changes into account. This ensures that all preference checks can be done very fast. Are you sure you want dynamic checks here?

File: src/plans/stopping_criterion.jl (hunk 10 above)
Comment (kellertuer) [reply src/plans/stopping_criterion.jl:637]: Oh, not sure how that happened ;)

File: tutorials/Optimize.qmd (hunk 12 above)
Comment (kellertuer) [reply tutorials/Optimize.qmd:184]: lang -> land

File: src/solvers/gradient_descent.jl (hunk 11 above)
Comment (kellertuer) [reply src/solvers/gradient_descent.jl:243]: Yes, I am not yet sure we might have more modes, but if it basically remains this one, then that function sounds fair.

File: src/plans/plan.jl (hunk 8 above)
Comment (kellertuer) [reply src/plans/plan.jl:83]: I wanted to avoid to have printed `"nothing"` in the string. That is what this avoids.

**Comment by kellertuer** [comment]
> The suggested usage of Preferences.jl is that you restart your Julia session to take any preference changes into account. This ensures that all preference checks can be done very fast. Are you sure you want dynamic checks here?

My honest answer is: I would prefer to be able to set this dynamically – otherwise I for example have no clue how to test that? How do I _restart Julia_ in testing to test the other mode?
For now this is persistent after starts, yes. But dynamic.

Besides the testing, I would also have not much clue how to implement what you refer to? So one would set the preference but that is never accessed but something different, because only on start (load of Manopt) the local preferences of preferences.jl are read? To me (but maybe I am just not that clever here) that reads like probably 100 more lines of code to split that into two layers, one that we load and one that we have during runtime?

File: src/plans/plan.jl (hunk 8 above)
Comment (kellertuer) [reply src/plans/plan.jl:83]: ah and if the value is length 0, eg for a string, this also avoids () and does not add brackets, like the last case does.

Comment (mateuszbaran) [reply src/plans/plan.jl:83]: I see. Though maybe it would be better to print strings in `"`, so that empty string is `""`?

Comment (kellertuer) [reply src/plans/plan.jl:83]: Hm, maybe, maybe we are also just overcomplicating things. In the very most cases this indeed will only be strings and `resetting to default ""` also does not read so nice.

**Comment by mateuszbaran** [comment]
I thought about it again and you are most likely right to keep preferences dynamic. Let's not over-complicate it.

File: src/solvers/gradient_descent.jl (hunk 11 above)
Comment (kellertuer) [reply src/solvers/gradient_descent.jl:243]: Will add the easy access method tomorrow. It will make a few places more readable.

File: docs/src/plans/index.md [hunk 13]
```diff
@@ -47,4 +47,10 @@ The column “generic” refers to a short hand that might be used for readabil
 Since the iterate is often stored in the states fields `s.p` one _could_ access the iterate
 often also with `:p` and similarly the gradient with `:X`.
 This is discouraged for both readability as well as to star more generic, and it is recommended
-to use `:Iterate` and `:Gradient` instead in generic settings.
\ No newline at end of file
+to use `:Iterate` and `:Gradient` instead in generic settings.
+
+YOu can further activate a “Tutorial” mode by `set_manopt_parameter!(:Mode, "Tutorial"). Internally, the following convenience function is available.
```
Comment (mateuszbaran) [inline docs/src/plans/index.md:52]:
```suggestion
You can further activate a “Tutorial” mode by `set_manopt_parameter!(:Mode, "Tutorial")`. Internally, the following convenience function is available.
```

File: src/plans/debug.jl [hunk 14]
```diff
@@ -946,7 +948,59 @@ function (d::DebugWarnIfFieldNotFinite)(
...
+end
+function (d::DebugWarnIfGradientNormTooLarge)(
+    mp::AbstractManoptProblem, st::AbstractManoptSolverState, i::Int
+)
+    if d.status !== :No
+        M = get_manifold(mp)
+        p = get_iterate(st)
+        X = get_gradient(st)
+        Xn = norm(M, p, X)
+        p_inj = d.factor * injectivity_radius(M, p)
```
Comment (mateuszbaran) [inline src/plans/debug.jl:989]: Given that tangent bundle has 0 injectivity radius, I'd suggest using `max_stepsize` here instead of `injectivity_radius`.

File: tutorials/Optimize.qmd [hunk 15]
```diff
@@ -156,9 +155,62 @@ Then we reach approximately the same point as in the previous run, but in far le
...
+grad_f2(M, p) = sum(grad_distance.(Ref(M), data, Ref(p)));
+```
+
+And we consider the gradient at the starting point in norm
+
+```{julia}
+norm(M, data[1], grad_f2(M, data[1]))
+```
+
+On the sphere, when we follow a geodesic, we “return” to the start point after length $2π$. If we “land” short before the starting point due to a gradient of length just shy of $2π$, line search would not make much sense probably.
```
Comment (mateuszbaran) [inline tutorials/Optimize.qmd:184]: Line search will always return something but it won't take into account that gradient is just a local approximation of the objective.

I think it would be better to rephrase it so that you don't have to use the word "probably".

File: src/plans/plan.jl [hunk 16]
```diff
@@ -35,6 +37,64 @@ function get_manopt_parameter(f, e::Symbol, args...)
 end
 get_manopt_parameter(f, args...) = nothing
 
+"""
+    get_manopt_parameter(element::Symbol; default=nothing)
+
+Access global [`Manopt`](@ref) parametersadressed by a symbol `element`.
+We first dispatch on the value of `element`.
+
+if the value is not set, `default` is returned
```
Comment (mateuszbaran) [inline src/plans/plan.jl:46]:
```suggestion
If the value is not set, `default` is returned.
```

**Review by mateuszbaran — COMMENTED** [review]
Just a few last minor things.

File: docs/src/plans/index.md (hunk 13 above)
Comment (kellertuer) [reply docs/src/plans/index.md:52]: Good that I am starting to use vale, that also reports on these. Somehow my VS Code Spellchecker does not want to work as I think it should.

File: src/plans/debug.jl (hunk 14 above)
Comment (kellertuer) [reply src/plans/debug.jl:989]: I think I do not understand what you mean here – I think the tangent bundle has even infectivity of infinite, since in every tangent space we have a vector space?

Besides that, you are right, maybe max_stepsize fits a bit better.

File: tutorials/Optimize.qmd (hunk 15 above)
Comment (kellertuer) [reply tutorials/Optimize.qmd:184]: You are right, it will always return something, but in the situation described, the step size might actually just be 0, since for such a long vector -grad is basically a very scaled down +grad .

I will try to rephrase this to avoid the probably though.

File: src/plans/debug.jl (hunk 14 above)
Comment (kellertuer) [reply src/plans/debug.jl:989]: Changed it to `max_stepsize`.

Comment (mateuszbaran) [reply src/plans/debug.jl]: Tangent *space* has infinite injectivity radius but tangent *bundle* (with the Sasaki metric) has injectivity radius 0 if the underlying manifold is not flat: https://mathoverflow.net/a/94348 .

**Review by mateuszbaran — APPROVED** [review]

File: src/plans/debug.jl (hunk 14 above)
Comment (kellertuer) [reply src/plans/debug.jl]: Sasaki Metric was the thing I missed here.

**Comment by kellertuer** [comment]
GitHub has some Hickups today, first not being able to pull from gh-pages, now this is merged but this PR not closed?! Strange.

**Comment by mateuszbaran** [comment]
I don't see this as merged, only closed.

**Comment by mateuszbaran** [comment]
Wait, master branch history shows it's merged but his PR does not. Weird.

**Comment by kellertuer** [comment]
I see a commit that refers to this on master, but also this not marked as merged. But if I open it, the changelog check fails, since the changelog is identical with master – so it _is_ on master. I am confused.
